### PR TITLE
Added an analyzer to flag stack only types on the heap

### PR DESCRIPTION
**Pull request by KrzysztofCwalina**
This analyzer will flag a field, if the type of the field is a stack only type, 
but the type containing the field is not stack-only.

Stack-only types are types that have an attribute with simple name equal to "StackOnlyAttribute" applied to them, e.g. 

using System;

class StackOnlyAttribute : Attribute {}

[StackOnly] 
struct StackOnlyStruct {}

class SmoeClass { 
    StackOnlyStruct _field;
}

The field "_field" will be flagged.

I am planing to use this analyzer with Span<T> to make it safe.

### CODE REVIEW

File: src/Microsoft.CodeAnalysis.StackOnlyTypes/Analyzer.Test/Verifiers/DiagnosticVerifier.cs [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+            var actualSpan = actual.GetLineSpan();
+
+            Assert.IsTrue(actualSpan.Path == expected.Path || (actualSpan.Path != null && actualSpan.Path.Contains("Test0.") && expected.Path.Contains("Test.")),
+                string.Format("Expected diagnostic to be in file \"{0}\" was actually in file \"{1}\"\r\n\r\nDiagnostic:\r\n    {2}\r\n",
+                    expected.Path, actualSpan.Path, FormatDiagnostics(analyzer, diagnostic)));
+
+            var actualLinePosition = actualSpan.StartLinePosition;
+
+            // Only check line position if there is an actual line in the real diagnostic
+            if (actualLinePosition.Line > 0)
```
Comment (pdelvo) [inline src/Microsoft.CodeAnalysis.StackOnlyTypes/Analyzer.Test/Verifiers/DiagnosticVerifier.cs]: I dont think that this is a good idea. See https://github.com/dotnet/roslyn/issues/1641

If an analyzer reports the diagnostic at line/column start there is no way for a unit test to catch that.

**Comment by KrzysztofCwalina** [comment]
@dotnet-bot test this please

**Comment by livarcocc** [comment]
It seems like you have some weird commits in your commits. There is a commit from me and another from Stephen.

**Comment by KrzysztofCwalina** [comment]
fixed the commit issue

File: src/Microsoft.CodeAnalysis.StackOnlyTypes/Analyzer.Test/Verifiers/DiagnosticVerifier.cs (hunk 1 above)
Comment (KrzysztofCwalina) [inline src/Microsoft.CodeAnalysis.StackOnlyTypes/Analyzer.Test/Verifiers/DiagnosticVerifier.cs]: Thanks. I removed both of the checks.

**Comment by omariom** [comment]
Storing Span in a field of a ref type is currently unsafe? Why?

**Comment by KrzysztofCwalina** [comment]
It's not safe because of so called "struct tearing" issue. Span<T> has two fields that cannot be updated atomically. When a Span filed is being assigned to on one thread, and another thread reads the field at the same time, the other thread might see inconsistent state (first filed update, but not the other). And because Span<T> operates on unsafe pointers, this in turn might lead to AVs or even memory corruption.

**Comment by omariom** [comment]
But readonly fields of reference types are safe in this sense, aren't they?

**Comment by KrzysztofCwalina** [comment]
Not sure if that's what you mean, but references are word size and so can be updated atomically.

**Comment by omariom** [comment]
I mean this case:
```C#
class Foo
{
    private readonly Span<byte> bar;
}
```

**Comment by KrzysztofCwalina** [comment]
Yeah, good point. I think it's fine. I should probably update the analyzer accordingly.

**Pull request merged**